File: templates/problems/15_greedy/03_jump_game_ii/solution.go
package jump_game_ii

// solveJump uses a greedy BFS approach. O(n) time, O(1) space.
func solveJump(nums []int) int {
	jumps := 0
	curEnd := 0
	farthest := 0
	for i := 0; i < len(nums)-1; i++ {
		if i+nums[i] > farthest {
			farthest = i + nums[i]
		}
		if i == curEnd {
			jumps++
			curEnd = farthest
		}
	}
	return jumps
}

File: templates/problems/15_greedy/03_jump_game_ii/typescript/solution.ts
/**
 * Greedy BFS approach. O(n) time, O(1) space.
 */
export function solveJump(nums: number[]): number {
    let jumps = 0;
    let curEnd = 0;
    let farthest = 0;
    for (let i = 0; i < nums.length - 1; i++) {
        if (i + nums[i] > farthest) farthest = i + nums[i];
        if (i === curEnd) {
            jumps++;
            curEnd = farthest;
        }
    }
    return jumps;
}
